src/scf/newton.jl

```diff
@@ -0,0 +1,317 @@
+    G_to_r(basis, δρ_fourier) ./ count
+end
+
+@views function apply_K(basis::PlaneWaveBasis, δψ, ψ, ρ, occupation)
+
+    δψ = proj_tangent(δψ, ψ)
+    δρ = compute_δρ(basis, ψ, δψ, occupation)
+    δV = apply_kernel(basis, δρ; ρ=ρ)
+
+    Kδψ = map(enumerate(ψ)) do (ik, ψk)
+        kpt = basis.kpoints[ik]
+        δVψk = similar(ψk)
+
+        for n = 1:size(ψk, 2)
+            ψnk_real = G_to_r(basis, kpt, ψk[:,n])
+            δVψnk_real = δV[:, :, :, kpt.spin] .* ψnk_real
+            δVψk[:, n] = r_to_G(basis, kpt, δVψnk_real)
+        end
+        δVψk
+    end
+    # ensure projection onto the tangent space
+    proj_tangent!(Kδψ, ψ)
+end
+
+function NewtonConvergenceDensity(tolerance)
+    info -> (norm(info.ρ_next - info.ρ) * sqrt(info.basis.dvol) < tolerance)
+end
+
+function NewtonDefaultCallback()
+    prev_energies = nothing
+    function callback(info)
+        !mpi_master() && return info  # Printing only on master
+        if info.stage == :finalize
+            info.converged || @warn "Newton not converged."
+            return info
+        end
+        collinear = info.basis.model.spin_polarization == :collinear
+
+        if info.n_iter == 1
+            E_label = haskey(info.energies, "Entropy") ? "Free energy" : "Energy"
+            magn    = collinear ? ("   Magnet", "   ------") : ("", "")
+            @printf "n     %-12s      Eₙ-Eₙ₋₁     ρ_next-ρ%s\n" E_label magn[1]
+            @printf "---   ---------------   ---------   --------%s\n" magn[2]
+        end
+        E    = isnothing(info.energies) ? Inf : info.energies.total
+        Δρ   = norm(info.ρ_next - info.ρ) * sqrt(info.basis.dvol)
+        if size(info.ρ_next, 4) == 1
+            magn = NaN
+        else
+            magn = sum(spin_density(info.ρ_next)) * info.basis.dvol
+        end
+
+        Estr   = (@sprintf "%+15.12f" round(E, sigdigits=13))[1:15]
+        prev_E = prev_energies === nothing ? Inf : prev_energies.total
+        ΔE     = prev_E == Inf ? "      NaN" : @sprintf "% 3.2e" E - prev_E
+        Mstr = collinear ? "   $((@sprintf "%6.3f" round(magn, sigdigits=4))[1:6])" : ""
+        @printf "% 3d   %s   %s   %2.2e%s\n" info.n_iter Estr ΔE Δρ Mstr
+        prev_energies = info.energies
+
+        flush(stdout)
+        info
+    end
+end
+
+"""
+    solve_ΩplusK(basis::PlaneWaveBasis{T}, ψ, res, occupation;
+                 tol_cg=1e-10, verbose=false) where T
+
+Return δψ where (Ω+K) δψ = rhs
+"""
+function solve_ΩplusK(basis::PlaneWaveBasis{T}, ψ, rhs, occupation;
+                     tol_cg=1e-10, verbose=false) where T
+
+    Nk = length(basis.kpoints)
+
+    # compute quantites at the point which define the tangent space
+    ρ = compute_density(basis, ψ, occupation)
+    _, H = energy_hamiltonian(basis, ψ, occupation; ρ=ρ)
+
+    # pack and unpack
+    pack(ψ) = pack_ψ(basis, ψ)
+    unpack(x) = unpack_ψ(basis, x)
+
+    # project rhs on the tangent space before starting
+    proj_tangent!(rhs, ψ)
+    rhs_pack = pack(rhs)
```

Comment:
rhs_pack = pack(proj_tangent!(rhs, ψ))

Reply:
really? I like having mutation in its own line (but it doesn't matter too much)

Reply:
I think that's why you have the `!`, no? But I agree it's a matter of style.

Reply:
! is easily lost if in an expression, whereas it stands out more on its own line